src/Voron/Impl/FreeSpace/StreamBitArray.cs

```diff
@@ -84,88 +86,177 @@ private void Write(byte* ptr)
+            var item = _inner[idx];
             return idx * 32 + BitOperations.TrailingZeroCount(item);
         }
         return -1;
     }
 
-    private bool TryScalarSearch<TModify>(int scalarSearch, int vectorStart, out int trailingZeroCount)
-        where TModify: struct, IModifyBuffer
+    public int FindRange(int num)
     {
-        for (int i = scalarSearch / 32; i < Vector256<int>.Count; i++)
+        return num switch
         {
-            var bitsToZero = scalarSearch % 32;
-            scalarSearch = 0;
-            var bits = default(TModify).Modify(_inner[vectorStart + i]) & (-1 << bitsToZero);
-            if (bits != 0)
-            {
-                trailingZeroCount = (vectorStart+i) * 32 + BitOperations.TrailingZeroCount(bits);
-                return true;
-            }
-        }
-
-        trailingZeroCount = -1;
-        return false;
+            >= 32 => FindRangeUsingWords(num),
+            _ => FindRangeUsingBits(num)
+        };
     }
-    
-    private struct Nothing : IModifyBuffer
-    {
-        public uint Modify(uint i) => i;
 
-        public Vector256<uint> Modify(Vector256<uint> v) => v;
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private bool FindRangeInVector(Vector256<uint> v, int num, out int bits)
+    {
+        // find consecutive range: https://stackoverflow.com/a/37903049/6366
+        // note that this is only _inside_ words
+        for (int j = 0; j < num-1; j++)
+        {
+            v &= v << 1;
+        }
+            
+        var gt = Vector256.GreaterThan(v, Vector256<uint>.Zero);
+        if (gt == Vector256<uint>.Zero)
+        {
+            bits = -1;
+            return false;
+        }
+            
+        var mask = gt.ExtractMostSignificantBits();
+        int vectorIndex = BitOperations.TrailingZeroCount(mask);
+        uint element = v.GetElement(vectorIndex);
+        var item = element - (element & (element - 1));
+        var bitPos = BitOperations.TrailingZeroCount(item) - (num - 1);
+        bits = bitPos + vectorIndex * 32;
+        return true;
     }
-    
-    private struct Inverse : IModifyBuffer
+
+    private int FindRangeUsingBits(int num)
     {
-        public uint Modify(uint i) => ~i;
+        ref uint shiftedInner = ref Unsafe.AddByteOffset(ref _inner[0], 2);
+        // we are skipping the first 2 bytes, and we are ***reading 2 bytes past the buffer***
+        // this is fine to do, since we know that after the buffer, we have the SetCount field
+        Debug.Assert(
+            Unsafe.AsPointer(ref SetCount) == Unsafe.AsPointer(ref _inner[CountOfItems]),
+            "We are intentionally reading past the end of the buffer"
+        );
+        for (int i = 0; i < CountOfItems; i += Vector256<uint>.Count)
+        {
+            var aligned = Vector256.LoadUnsafe(ref _inner[0], (nuint)i);
+            var unaligned = Vector256.LoadUnsafe(ref shiftedInner, (nuint)i);
+
```

Comment:
If we are sure this is aligned we could use `Vector256.LoadAligned` instead.

Reply:
No, this will require us to be aligned on 32 bytes boundary.
And we have no way to require that, I'm afraid.